crates/services/upgradable-executor/src/executor.rs

```diff
@@ -291,4 +404,407 @@ where
+
+            // Then
+            assert!(result.is_err());
+        }
+
+        fn storage_with_state_transition(
+            next_version: StateTransitionBytecodeVersion,
+        ) -> Storage {
+            let mut storage = storage();
+            let mut tx = storage.write_transaction();
+            tx.storage_as_mut::<StateTransitionBytecodeVersions>()
+                .insert(&next_version, WASM_BYTECODE)
+                .unwrap();
+            tx.commit().unwrap();
+
+            storage
+        }
+
+        #[test]
+        fn can_validate_block_with_next_version__native_strategy() {
+            // Given
+            let next_version = Executor::<Storage, DisabledRelayer>::VERSION + 1;
+            let storage = storage_with_state_transition(next_version);
+            let executor = Executor::native(storage, DisabledRelayer, Config::default());
+            let block = valid_block(next_version);
+
+            // When
+            let result = executor
+                .execute_without_commit(ExecutionTypes::Validation(block))
+                .map(|_| ());
+
+            // Then
+            assert_eq!(Ok(()), result);
+        }
+
+        #[test]
+        fn can_validate_block_with_next_version__wasm_strategy() {
+            // Given
+            let next_version = Executor::<Storage, DisabledRelayer>::VERSION + 1;
+            let storage = storage_with_state_transition(next_version);
+            let executor = Executor::wasm(storage, DisabledRelayer, Config::default());
+            let block = valid_block(next_version);
+
+            // When
+            let result = executor
+                .execute_without_commit(ExecutionTypes::Validation(block))
+                .map(|_| ());
+
+            // Then
+            assert_eq!(Ok(()), result);
+        }
+
+        // The test verifies that `Executor::get_module` method caches the compiled WASM module.
+        // If it doesn't cache the modules, the test will fail with a timeout.
+        #[test]
+        #[ntest::timeout(40_000)]
+        fn reuse_cached_compiled_module__native_strategy() {
+            // Given
+            let next_version = Executor::<Storage, DisabledRelayer>::VERSION + 1;
+            let storage = storage_with_state_transition(next_version);
+            let executor = Executor::native(storage, DisabledRelayer, Config::default());
+            let block = valid_block(next_version);
+
+            // When
+            for _ in 0..1000 {
+                let result = executor
+                    .execute_without_commit(ExecutionTypes::Validation(block.clone()))
+                    .map(|_| ());
+
+                // Then
+                assert_eq!(Ok(()), result);
+            }
+        }
+
+        // The test verifies that `Executor::get_module` method caches the compiled WASM module.
+        // If it doesn't cache the modules, the test will fail with a timeout.
+        #[test]
+        #[ntest::timeout(40_000)]
```

Comment:
nit: Potentially flaky, especially in the CI. Cannot thing of an easy way to do this better, though.